packages/webapp/src/components/Task/AddSoilAmendmentProducts/QuantityApplicationRate/useQuantityApplicationRate.ts

```diff
@@ -0,0 +1,172 @@
+    return isNaN(Number(volumeValue));
+  });
+
+  const toggleMeasure = () => {
+    setIsWeight((prev) => !prev);
+  };
+
+  /* Update application area + rate based on percent of location */
+  const onPercentLocationChange = (percent_of_location: number) => {
+    const calculatedArea = (total_area * Math.min(percent_of_location || 100, 100)) / 100;
+    setValue(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED, calculatedArea);
+
+    /* set unit of the total area component according to default breakpoints */
+    setValue(
+      TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED_UNIT,
+      /* @ts-ignore */
+      getUnitOptionMap()[getDefaultUnit(location_area, calculatedArea, system).displayUnit],
+      { shouldValidate: true },
+    );
+
+    updateApplicationRate();
+  };
+
+  const updateApplicationRate = () => {
+    const weight = getValues(TASK_PRODUCT_FIELD_NAMES.WEIGHT);
+    const application_area = getValues(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED);
+    const volume = getValues(TASK_PRODUCT_FIELD_NAMES.VOLUME);
+    if ((weight || weight === 0) && isWeight && weight_unit && application_area) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT,
+        convert(weight / application_area)
+          .from('kg/m2') // database unit weight / database unit area
+          .to('kg/ha'), // database unit application_rate_weight
+        { shouldValidate: true },
+      );
+    } else if ((volume || volume === 0) && !isWeight && volume_unit && application_area) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME,
+        convert(volume / application_area)
+          .from('l/m2') // database unit volume / database unit area
+          .to('l/ha'), // database unit application_rate_volume
+        { shouldValidate: true },
+      );
+    }
+  };
+
+  const updateQuantity = () => {
+    const application_rate_weight = getValues(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_WEIGHT);
+    const application_rate_volume = getValues(TASK_PRODUCT_FIELD_NAMES.APPLICATION_RATE_VOLUME);
+    const application_area = getValues(TASK_PRODUCT_FIELD_NAMES.TOTAL_AREA_AMENDED);
+    if (
+      isWeight &&
+      (application_rate_weight || application_rate_weight === 0) &&
+      application_rate_weight_unit &&
+      application_area
+    ) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.WEIGHT,
+        convert(application_rate_weight).from('kg/ha').to('kg/m2') * application_area,
+        { shouldValidate: true },
+      );
+    } else if (
+      !isWeight &&
+      (application_rate_volume || application_rate_volume === 0) &&
+      application_rate_volume_unit &&
+      application_area
+    ) {
+      setValue(
+        TASK_PRODUCT_FIELD_NAMES.VOLUME,
+        convert(application_rate_volume).from('l/ha').to('l/m2') * application_area,
+        { shouldValidate: true },
+      );
+    }
+  };
+
+  /* Trigger recalculation when units are changed */
+  useEffect(() => {
```

Comment:
Isn't it possible to do this in the change handler for Unit? 

`<Unit  onChangeUnitOption={() => updateApplicationRate()} />`

If it works then I think its cleaner and potentially more performant than useEffect. Same with quantity update below.